### ISSUE TITLE
tests/posix/common: test_posix_pthread_execution() reboots repeatedly for portability.posix

### ISSUE THREAD

**Issue by cinlyooi-intel** [zephyrproject-rtos/zephyr#15247]
**Describe the bug**
test_posix_phread_execution()reboots repeatedly.

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=frdm_k64f
3. make
4. make run

**Screenshots or console output**
```
***** ***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.14.0-rc3-148-g318a98a7fb1f (delayed boot 1000ms) *****
Running test suite posix_apis
===================================================================
starting test - test_posix_pthread_execution
Thread 0 starting with scheduling policy 0 & priority 15
Thread 1 starting with scheduling policy 0 & priority 15
Thread 2 starting with scheduling policy 0 & priority 15
***** delaying boot 1000ms (per build configuration) *****
```

**Environment (please complete the following information):**
 - OS: fedora
 - Toolchain : zephyr
 - Commit SHA or Version used: 318a98a7fb1f604369ff1f2ee99608836176a0d4

**Comment by andrewboie** [comment]
that test sits for a few seconds after those printouts (it's doing a lot of context switch thrashing between threads) and then continues...could this be a watchdog reset?

**Comment by MaureenHelm** [comment]
I don't think this is a watchdog reset. The watchdog gets disabled during boot and doesn't get re-enabled unless you enable the watchdog driver (which we don't by default).

We are getting a usage fault and git bisect is pointing to:
```
commit 7df0216d1e2a2800225d2d3756f5decb0568188d
Author: Andy Ross <andrew.j.ross@intel.com>
Date:   Wed Jul 25 13:55:59 2018 -0700

    kernel/mutex: Spinlockify
    
    Use a subsystem lock, not a per-object lock.  Really we want to lock
    at mutex granularity where possible, but (1) that has non-trivial
    memory overhead vs. e.g. directly spinning on the mutex state and (2)
    the locking in a few places was originally designed to protect access
    to the mutex *owner* priority, which is not 1:1 with a single mutex.
    
    Basically the priority-inheriting mutex code will need some rework
    before it works as a fine-grained locking abstraction in SMP.
    
    Note that this fixes an invisible bug: with the older code,
    k_mutex_unlock() would actually call irq_unlock() twice along the path
    where there was a new owner, which is benign on existing architectures
    (so long as the key argument is unchanged) but was never guaranteed to
    work.  With a spinlock, unlocking an unlocked/unowned lock is a
    detectable assertion condition.
    
    Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
```
cc: @andyross

**Comment by MaureenHelm** [comment]
I think this is related to #15411. The same suspicious commit 2895da02a44980e901fbcdd7ec87edc2016605c5 changes this test from failing to passing!

**Comment by andrewboie** [comment]
@MaureenHelm I think you may be onto something. For unrelated reasons I was looking into the implementations of tests/kernel/critical and tests/posix/common for #14642 . These two tests have something in common: they both do a crazy amount of context switch thrashing between various threads.

So if we have some kind of race condition involved with context switching, it's reasonable for both these tests to encounter problems.

**Comment by nashif** [comment]
this bug however was opened before 2895da0 was committed...

**Comment by MaureenHelm** [comment]
> this bug however was opened before [2895da0](https://github.com/zephyrproject-rtos/zephyr/commit/2895da02a44980e901fbcdd7ec87edc2016605c5) was committed...

The test fails before that commit and passes after it.

**Comment by nashif** [comment]
> The test fails before that commit and passes after it.

oh, I misread the comment above

**Comment by ioannisg** [comment]
This makes me think the root-cause of these fails might be on kernel - not on the timer driver side.

**Comment by nashif** [comment]
the test fails if you add CONFIG_BOOT_DELAY=1000

**Comment by andyross** [comment]
This board has a watchdog, right?  Is it just that the boot delay is happening with interrupts locked and/or in a state where the watchdog hasn't been set up for the right timeout yet?

**Comment by MaureenHelm** [comment]
The watchdog gets disabled very early in the boot process before the delay, and doesn't get reenabled unless `CONFIG_WATCHDOG=y`. I'm also able to reproduce this without the boot delay.
